### PR TITLE
Add omitempty to dnsmanagementpolicy

### PR DESCRIPTION
**Pull request by thejasn**
Signed-off-by: Thejas N <thn@redhat.com>

### CODE REVIEW

**Comment by JoelSpeed** [comment]
/lgtm
/assign @deads2k for approval

This was discussed on slack, the lack of omitempty causes the default to be set by openapi which then won't pass validation. With omitempty, validation passes and then the default is set

File: openapi/openapi.json [hunk 1]
```diff
@@ -23722,8 +23722,7 @@
       "properties": {
         "dnsManagementPolicy": {
           "description": "dnsManagementPolicy indicates if the lifecycle of the wildcard DNS record associated with the load balancer service will be managed by the ingress operator. It defaults to Managed. Valid values are: Managed and Unmanaged.",
-          "type": "string",
-          "default": ""
```
Comment (deads2k) [inline openapi/openapi.json:23726]: So the problem is that this value here should be Managed?

Comment (JoelSpeed) [reply openapi/openapi.json:23726]: Yes, we need the defaults to be accepted in both OpenAPI and controller-tools, except they use different tags, so right now the default is `""` which is causing validation issues. You must have omitempty to have a non-empty default here as well

File: operator/v1/types_ingress.go [hunk 2]
```diff
@@ -393,7 +393,7 @@ type LoadBalancerStrategy struct {
 	// +kubebuilder:default:="Managed"
 	// +kubebuilder:validation:Optional
 	// +optional
-	DNSManagementPolicy LoadBalancerDNSManagementPolicy `json:"dnsManagementPolicy"`
+	DNSManagementPolicy LoadBalancerDNSManagementPolicy `json:"dnsManagementPolicy,omitempty"`
```
Comment (JoelSpeed) [inline operator/v1/types_ingress.go:396]:
```suggestion
	// +default="Managed"
	DNSManagementPolicy LoadBalancerDNSManagementPolicy `json:"dnsManagementPolicy,omitempty"`
```

File: operatoringress/v1/types.go [hunk 3]
```diff
@@ -72,7 +72,7 @@ type DNSRecordSpec struct {
 	// +kubebuilder:default:="Managed"
 	// +kubebuilder:validation:Optional
 	// +optional
-	DNSManagementPolicy DNSManagementPolicy `json:"dnsManagementPolicy"`
+	DNSManagementPolicy DNSManagementPolicy `json:"dnsManagementPolicy,omitempty"`
```
Comment (JoelSpeed) [inline operatoringress/v1/types.go:75]:
```suggestion
	// +default="Managed"
	DNSManagementPolicy LoadBalancerDNSManagementPolicy `json:"dnsManagementPolicy,omitempty"`
```

**Review by JoelSpeed — COMMENTED** [review]
@thejasn Please apply the two suggestions and regenerate and I think this is good

**Comment by JoelSpeed** [comment]
I've been experimenting with this and as far as I can tell this makes things works as they stand today with client-side and server-side validation

We want to have the correct default in the openapi.json, and to achieve this, we need to add `omitempty` and the default tag.

There's an open question about `optional` vs `required` for defaulting but I think that can be handled separately as this resolves the validations for server side (which means anything using client-go should work) and client-side validation is also happy right now.

/lgtm

File: operator/v1/types_ingress.go [hunk 4]
```diff
@@ -393,7 +393,8 @@ type LoadBalancerStrategy struct {
 	// +kubebuilder:default:="Managed"
 	// +kubebuilder:validation:Optional
 	// +optional
```
Comment (deads2k) [inline operator/v1/types_ingress.go:395]: required.

File: operator/v1/types_ingress.go [hunk 5]
```diff
@@ -393,7 +393,8 @@ type LoadBalancerStrategy struct {
...
-	DNSManagementPolicy LoadBalancerDNSManagementPolicy `json:"dnsManagementPolicy"`
+	// +default="Managed"
+	DNSManagementPolicy LoadBalancerDNSManagementPolicy `json:"dnsManagementPolicy,omitempty"`
```
Comment (deads2k) [inline operator/v1/types_ingress.go:397]: and leaving omitempty I'll deal with for now.

File: operator/v1/types_ingress.go (hunk 4 above)
Comment (deads2k) [reply operator/v1/types_ingress.go:395]: > required.

not optional anyway.

**Comment by thejasn** [comment]
/hold 
cc: @Miciah

File: operator/v1/types_ingress.go [hunk 6]
```diff
@@ -391,9 +391,10 @@ type LoadBalancerStrategy struct {
 	// Valid values are: Managed and Unmanaged.
 	//
 	// +kubebuilder:default:="Managed"
-	// +kubebuilder:validation:Optional
-	// +optional
-	DNSManagementPolicy LoadBalancerDNSManagementPolicy `json:"dnsManagementPolicy"`
+	// +kubebuilder:validation:Required
+	// +required
```
Comment (JoelSpeed) [inline operator/v1/types_ingress.go:395]: Delete this line please, it doesn't serve any purpose as it isn't picked up by any generators. `+kubebuilder:validation:Required` is the correct tag

File: operatoringress/v1/types.go [hunk 7]
```diff
@@ -70,9 +70,10 @@ type DNSRecordSpec struct {
 	// "Unmanaged".
 	//
 	// +kubebuilder:default:="Managed"
-	// +kubebuilder:validation:Optional
-	// +optional
-	DNSManagementPolicy DNSManagementPolicy `json:"dnsManagementPolicy"`
+	// +kubebuilder:validation:Required
+	// +required
```
Comment (JoelSpeed) [inline operatoringress/v1/types.go:74]: Delete this line please, it doesn't serve any purpose as it isn't picked up by any generators. `+kubebuilder:validation:Required` is the correct tag

**Pull request merged**